include/volume/volume_cooling_gaussians_crhmc.hpp

```diff
@@ -0,0 +1,448 @@
+    const unsigned int totalSteps = ((int)150/((1.0 - frac) * error))+1;
+
+    if (a_vals[0]<a_stop) a_vals[0] = a_stop;
+
+#ifdef VOLESTI_DEBUG
+    std::cout<<"first gaussian computed "<< a_vals[0] <<std::endl;
+#endif
+
+#ifdef VOLESTI_DEBUG
+    std::cout<<"Computing the sequence of gaussians..\n"<<std::endl;
+#endif
+
+    while (true)
+    {
+
+        NT curr_fn = 0;
+        NT curr_its = 0;
+        auto steps = totalSteps;
+
+        
+        //create the crhmc problem for this variance
+        int dimension = P.dimension();
+        func_params f_params = func_params(Point(dimension), a_vals[it], 1);
+        
+        Func f(f_params);
+        Grad g(f_params);
+        Hess h(f_params);
+
+        Input input = convert2crhmc_input<Input, Polytope, Func, Grad, Hess>(P, f, g, h);
+
+        typedef crhmc_problem<Point, Input> CrhmcProblem;
+        CrhmcProblem problem = CrhmcProblem(input);
+
+        Point p = Point(problem.center);
+
+        if(problem.terminate){return;}
+
+        problem.options.simdLen = simdLen;
+        walk_params params(input.df, p.dimension(), problem.options);
+
+        if (input.df.params.eta > 0) {
+            params.eta = input.df.params.eta;
+        }
+
+        int dim;
+        dim = p.dimension();
+
+        //create the walk object for this problem
+        WalkType walk = WalkType(problem, p, input.df, input.f, params);
+
+        // Compute the next gaussian
+        NT next_a = get_next_gaussian<WalkType, walk_params, RandomPointGenerator, simdLen, Grad, Func, CrhmcProblem>
+                      (P, p, a_vals[it], N, ratio, C, walk_length, rng, g, f, params, problem, walk);
+
+#ifdef VOLESTI_DEBUG
+    std::cout<<"Next Gaussian " << next_a <<std::endl;
+#endif
+
+        // Compute some ratios to decide if this is the last gaussian
+        for (unsigned  int j = 0; j < steps; j++)
+        {
+            walk.template apply(rng, walk_length);
+            p = walk.getPoint();
+#ifdef VOLESTI_DEBUG
+    std::cout<<"Walk point " << std::endl;
+    p.print();
+#endif
+            curr_its += 1.0;
+            curr_fn += eval_exp(p, next_a) / eval_exp(p, a_vals[it]);
+        }
+
+#ifdef VOLESTI_DEBUG
+    std::cout<<"Condition function ratio " << curr_fn/curr_its <<std::endl;
+#endif
+
+        // Remove the last gaussian.
+        // Set the last a_i equal to zero
+        if (next_a>0 && curr_fn/curr_its>(1.0+tol))
+        {
+            a_vals.push_back(next_a);
+            it++;
+        } else if (next_a <= 0)
+        {
+            a_vals.push_back(a_stop);
+            it++;
+            break;
+        } else {
+            a_vals[it] = a_stop;
+            break;
+        }
+    }
+#ifdef VOLESTI_DEBUG
+    std::cout<<"first gaussian after WHILE "<< a_vals[0] <<std::endl;
+#endif
+}
+
+template
+<
+    typename Polytope,
+    typename RandomNumberGenerator,
+    typename WalkTypePolicy = CRHMCWalk,
+    int simdLen = 8
+>
+double volume_cooling_gaussians_crhmc(Polytope& Pin,
```

Comment:
could we keep the same name? It can be overloaded right?

Reply:
Another question, what are the changes between this and the original volume_cooling_gaussians? If the changes are only the calls to get_first_gaussian and get_annealing_schedule and the type of the walk, is it possible to use the same function to avoid duplication of the code?

Reply:
Actually not really. There are slight changes especially to the walk call, which is done within this function as well. For example walk instantiation, apply and after we apply the walk, in the current crhmc implementation the point is not automatically updated, so we need to use p = walk.getPoint();